drivers/flash/flash_bflb.c

```diff
@@ -0,0 +1,1564 @@
+static int flash_bflb_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct flash_bflb_data *data = dev->data;
+	unsigned int	locker;
+	int ret;
+
+	if (length == 0) {
+		return 0;
+	}
+
+	ret = flash_bflb_is_valid_range(address, length);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (flash_bflb_is_in_xip(&flash_bflb_read)) {
+		return -ENOTSUP;
+	}
+
+	/* interrupting would break, likely to access XIP*/
+	locker = irq_lock();
+
+	ret = flash_bflb_save_xip_state(dev);
+	if (ret != 0) {
+		irq_unlock(locker);
+		return ret;
+	}
+
+	ret = flash_bflb_read_sahb_do(data, address, buffer, length);
+
+	if (ret != 0) {
+		flash_bflb_restore_xip_state(data);
+	} else {
+		ret = flash_bflb_restore_xip_state(data);
+	}
+	irq_unlock(locker);
+
+	return ret;
+}
+
+#else
+
+/* copies flash data using XIP access */
+static int flash_bflb_read(const struct device *dev, off_t address, void *buffer, size_t length)
+{
+	struct flash_bflb_data *data = dev->data;
+	uint32_t	img_offset;
+	unsigned int	locker;
+	int ret;
+
+	if (length == 0) {
+		return 0;
+	}
+
+	ret = flash_bflb_is_valid_range(address, length);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (flash_bflb_is_in_xip(&flash_bflb_read)) {
+		return -ENOTSUP;
+	}
+
+	/* interrupting would break, likely to access XIP*/
+	locker = irq_lock();
+
+	/* get XIP offset / where code really is in flash, usually 0x2000 */
+	img_offset = flash_bflb_get_offset(data->reg_copy);
+
+	/* need set offset to 0 to access? */
+	if (address < img_offset) {
+
+		sys_cache_data_flush_and_invd_all();
+
+		/* set offset to 0 to access first (likely)0x2000 of flash */
+		flash_bflb_set_offset(data->reg_copy, 0);
+
+		/* copy data we need */
+		flash_bflb_xip_memcpy((uint8_t *)(address + BFLB_XIP_BASE),
+				      (uint8_t *)buffer, length);
+
+		sys_cache_data_flush_and_invd_all();
+
+		flash_bflb_set_offset(data->reg_copy, img_offset);
+	} else {
+		/* copy data we need */
+		flash_bflb_xip_memcpy((uint8_t *)(address + BFLB_XIP_BASE - img_offset),
+				      (uint8_t *)buffer, length);
+	}
+
+	/* done with interrupt breaking stuffs */
+	irq_unlock(locker);
+
+	return 0;
+}
+
+#endif
+
+static int flash_bflb_write(const struct device *dev,
+			     off_t address,
+			     const void *buffer,
+			     size_t length)
+{
+	struct flash_bflb_data *data = dev->data;
+	uint32_t	tmp;
+	unsigned int	locker;
+	int		ret;
+	uint32_t	cur_len, i;
+	struct bflb_flash_command write_cmd = {0};
+
+	if (length == 0) {
+		return 0;
+	}
+
+	ret = flash_bflb_is_valid_range(address, length);
+	if (ret != 0) {
+		return ret;
+	}
+
+	if (flash_bflb_is_in_xip(&flash_bflb_write)) {
+		return -ENOTSUP;
+	}
+
+	/* interrupting would break, likely to access XIP*/
+	locker = irq_lock();
+
+	ret = flash_bflb_save_xip_state(dev);
+	if (ret != 0) {
+		irq_unlock(locker);
+		return ret;
+	}
+
+	/* Check if the flash chip is OK to write to
+	 * TODO: Ability to disable this and similar when used with things that arent flashes
+	 */
```

Comment:
Remember that and log enhancement, it looks like personal note.

Reply:
Also for a future feature update, it sounds like you would want those removed